addons/stock_account/models/stock.py

```diff
@@ -270,32 +264,30 @@ def _get_accounting_data_for_valuation(self):
             raise UserError(_('You don\'t have any stock valuation account defined on your product category. You must define one before processing this operation.'))
         journal_id = accounts_data['stock_journal'].id
         return journal_id, acc_src, acc_dest, acc_valuation
-
+    
     def _prepare_account_move_line(self, qty, cost, credit_account_id, debit_account_id):
         """
         Generate the account.move.line values to post to track the stock valuation difference due to the
         processing of the given quant.
         """
         self.ensure_one()
-
+    
         if self._context.get('force_valuation_amount'):
             valuation_amount = self._context.get('force_valuation_amount')
         else:
-            if self.product_id.cost_method == 'average':
-                valuation_amount = cost if self.location_id.usage == 'supplier' and self.location_dest_id.usage == 'internal' else self.product_id.standard_price
-            else:
-                valuation_amount = cost if self.product_id.cost_method == 'real' else self.product_id.standard_price
+            valuation_amount = cost
+
         # the standard_price of the product may be in another decimal precision, or not compatible with the coinage of
         # the company currency... so we need to use round() before creating the accounting entries.
-        debit_value = self.company_id.currency_id.round(valuation_amount * qty)
-
+        debit_value = self.company_id.currency_id.round(valuation_amount)
+    
         # check that all data is correct
         if self.company_id.currency_id.is_zero(debit_value):
             if self.product_id.cost_method == 'standard':
                 raise UserError(_("The found valuation amount for product %s is zero. Which means there is probably a configuration error. Check the costing method and the standard price") % (self.product_id.name,))
             return []
         credit_value = debit_value
-
+    
         if self.product_id.cost_method == 'average' and self.company_id.anglo_saxon_accounting:
             # in case of a supplier return in anglo saxon mode, for products in average costing method, the stock_input
             # account books the real purchase price, while the stock account books the average price. The difference is
```

Comment:
@hbto Hello, thanks a lot for the insightful comments. I'm not familiar with the concept of kardex (maybe @jco-odoo is?) but i'm very interested by your AVCO examples. As you can see, the current implementation is immature and we have to deal with an added complexity, the edition of done move that could trigger a "replay" of the computing of the AVCO/FIFO chain. Most likely, this implementation will be merged as it in the saas-17 branch (in order to actually release it, we're two weeks late) but we'll fix it before any client will be able to use it.

We'll have a look at your examples and will try to add them to our test suite ([which is really poor at the moment](https://github.com/odoo-dev/odoo/blob/master-stock/addons/stock_account/tests/test_stockvaluation.py)). We'll get back to you in the next weeks about these usecases if we do not understand them correctly.

Regards